keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,854 @@
+   1. keeping an in-memory cache of how many users are on each shard (weighting "premium" users higher)
+   2. finding the `N` shards with the lowest user-load:
+       - if multiple shards have the same weighting, return them as "tied pods"
+       - if we find `N` empty shards, we can exit early, and return those as "chosen pods"
+   3. returning these lists of "chosen pods" and "tied pods"
+
+#### Story 3
+
+As a Data Engineer, I want to run an [Apache Spark](https://github.com/apache/spark) cluster and autoscale the number of workers
+while impacting the fewest running tasks on downscale, this allows me to save money by not under-utilizing Nodes,
+and reduce wasted time by minimising how many tasks are impacted by scaling.
+
+Solution:
+
+- (Similar to Story 1)
+
+#### Story 4
+
+As a Site Reliability Engineer, I want to ensure my NodeJS website maintains regional distribution when downscaling the
+number of replicas, this allows me to ensure uptime when a region experiences an outage.
+
+Solution:
+
+- I can run my NodeJS application in a Deployment.
+- I can use a HorizontalPodAutoscaler with CPU metrics to control how many `replicas` the Deployment has.
+- I can create a REST API (with TypeScript) for `downscalePodPicker` that runs in a Deployment, and answers a request to choose `N` Pods by:
+   1. keeping track of which region each Pod is in
+   2. finding `N` Pods that we can remove without violating the regional distribution requirements:
+       - if there are multiple acceptable options, we could choose the `N` Pods which are doing the least work
+   1. return these Pods as the "chosen pods"
+
+### Notes/Constraints/Caveats (Optional)
+
+- The `downscalePodPicker` APIs are written by users to suite their specific needs, and are not part of Kubernetes.
+- There are very few constraints on what a `downscalePodPicker` APIs can be, for example:
+   - they may be fully external to the Kubernetes cluster (e.g. running on an external metrics system)
+   - they may be part of the application which it controls (e.g. a REST API exposed by the application itself)
+   - they may be an adjacent Deployment running on the same Kubernetes cluster (e.g. a Python flask app)
+
+### Risks and Mitigations
+
+- A slow `downscalePodPicker` API may increase the time to downscale a ReplicaSet
+   - MITIGATION: we will provide a `timeoutSeconds` config, if the API has not successfully returned within this time, 
+     we proceed as if all Pods were "tied", and use the existing process to resolve the tie
+- The `downscalePodPicker` may not be reachable at all times, or may return an invalid response
+   - MITIGATION: we will provide a `maxRetries` config, if the API continues to fail after this many attempts, 
+     we proceed as if all Pods were "tied", and use the existing process to resolve the tie
+- Users may attempt to use `downscalePodPicker` and `controller.kubernetes.io/pod-deletion-cost` at the same time, resulting in unexpected behaviour
```

Comment:
how are you going to validate this? The design below seems to allow both.

Reply:
Are you talking about using `downscalePodPicker` and `controller.kubernetes.io/pod-deletion-cost` at the same time? If so, it's not that it will break something, it's just that the behavior is unexpected (because `controller.kubernetes.io/pod-deletion-cost` takes precedence over `downscalePodPicker`).

The mitigation is just to stress that using both features is not recommended.